src/librustc_privacy/lib.rs

```diff
@@ -20,22 +20,260 @@ use rustc::hir::intravisit::{self, Visitor, NestedVisitorMap};
+        self.skeleton().visit_trait(trait_ref)
+    }
+    fn visit_predicates(&mut self, predicates: Lrc<ty::GenericPredicates<'tcx>>) -> bool {
+        self.skeleton().visit_predicates(predicates)
+    }
+}
+
+struct DefIdVisitorSkeleton<'v, 'a, 'tcx, V>
+    where V: DefIdVisitor<'a, 'tcx> + ?Sized
+{
+    def_id_visitor: &'v mut V,
+    visited_opaque_tys: FxHashSet<DefId>,
+    dummy: PhantomData<TyCtxt<'a, 'tcx, 'tcx>>,
+}
+
+impl<'a, 'tcx, V> DefIdVisitorSkeleton<'_, 'a, 'tcx, V>
+    where V: DefIdVisitor<'a, 'tcx> + ?Sized
+{
+    fn visit_trait(&mut self, trait_ref: TraitRef<'tcx>) -> bool {
+        self.def_id_visitor.visit_def_id(trait_ref.def_id, "trait", &trait_ref) ||
+        trait_ref.visit_with(self)
+    }
+
+    fn visit_predicates(&mut self, predicates: Lrc<ty::GenericPredicates<'tcx>>) -> bool {
+        for (predicate, _) in &predicates.predicates {
+            let trait_ref = match predicate {
+                ty::Predicate::Trait(poly_predicate) => {
+                    poly_predicate.skip_binder().trait_ref
+                }
+                ty::Predicate::Projection(poly_predicate) => {
+                    if poly_predicate.skip_binder().ty.visit_with(self) {
+                        return true;
+                    }
+                    poly_predicate.skip_binder().projection_ty.trait_ref(self.def_id_visitor.tcx())
+                }
+                ty::Predicate::TypeOutlives(poly_predicate) => {
+                    if poly_predicate.skip_binder().0.visit_with(self) {
+                        return true;
+                    }
+                    continue;
+                }
+                ty::Predicate::RegionOutlives(..) => continue,
+                _ => bug!("unexpected predicate: {:?}", predicate),
+            };
+            if self.visit_trait(trait_ref) {
+                return true;
+            }
+        }
+        false
+    }
+}
+
+impl<'a, 'tcx, V> TypeVisitor<'tcx> for DefIdVisitorSkeleton<'_, 'a, 'tcx, V>
+    where V: DefIdVisitor<'a, 'tcx> + ?Sized
+{
+    fn visit_ty(&mut self, ty: Ty<'tcx>) -> bool {
+        let tcx = self.def_id_visitor.tcx();
+        match ty.sty {
+            ty::Adt(&ty::AdtDef { did: def_id, .. }, ..) |
+            ty::Foreign(def_id) |
+            ty::FnDef(def_id, ..) |
+            ty::Closure(def_id, ..) |
+            ty::Generator(def_id, ..) => {
+                if self.def_id_visitor.visit_def_id(def_id, "type", ty) {
+                    return true;
+                }
+                // Default type visitor doesn't visit fn signatures.
+                if let ty::FnDef(..) = ty.sty {
+                    if tcx.fn_sig(def_id).visit_with(self) {
+                        return true;
+                    }
+                }
+                // Inherent static methods don't have self type in substs,
+                // we have to check it additionally.
+                if let Some(assoc_item) = tcx.opt_associated_item(def_id) {
```

Comment:
also for this (e.g., for this part:)

```rust
/// The privacy def-ids for a type are defined recursively as:
/// 1. The `DefId`s syntactically within the type.
/// 2. For a `ty::FnDef`, the privacy def-ids within its
/// signature (TODO: why?)
/// <insert example here>
/// 3. For a `ty::FnDef` representing a function from an
/// (inherent) impl, the privacy def-ids within that impl's
/// trait-ref.
/// <insert example here>, but say, for:
/// ```
///     impl<T> Foo<T> {
///         fn bar() {}
///     }
/// ```
///
/// For `<Foo<Abc>>::bar`, we would include the def-ids
/// `Foo` and `Abc`.
```